codejail/jail.py

```diff
@@ -0,0 +1,329 @@
+
+        response = self.jail_code(
+            code=jailed_code,
+            stdin=stdin,
+            files=files,
+            slug=slug,
+            extra_files=extra_files,
+        )
+        if response.status != 0:
+            raise SafeExecException((
+                "Couldn't execute jailed code: stdout: {res.stdout!r}, "
+                "stderr: {res.stderr!r} with status code: {res.status}"
+            ).format(res=response))
+        globals_dict.update(json.loads(response.stdout))
+
+    def jail_code(self, code=None, files=None, extra_files=None, argv=None, stdin=None, slug=None):
+        """
+        Run code in the current jail.
+
+        `code` is a string containing the code to run.  If no code is supplied,
+        then the code to run must be in one of the `files` copied, and must be
+        named in the `argv` list.
+
+        `files` is a list of file paths, they are all copied to the jailed
+        directory.  Note that no check is made here that the files don't contain
+        sensitive information.  The caller must somehow determine whether to allow
+        the code access to the files.  Symlinks will be copied as symlinks.  If the
+        linked-to file is not accessible to the sandbox, the symlink will be
+        unreadable as well.
+
+        `extra_files` is a list of pairs, each pair is a filename and a bytestring
+        of contents to write into that file.  These files will be created in the
+        temp directory and cleaned up automatically.  No subdirectories are
+        supported in the filename.
+
+        `argv` is the command-line arguments to supply.
+
+        `stdin` is a string, the data to provide as the stdin for the process.
+
+        `slug` is an arbitrary string, a description that's meaningful to the
+        caller, that will be used in log messages.
+
+        Return a JailResult namedtuple with:
+
+            .status: exit status of the process: an int, 0 for success
+            .stdout: stdout of the program, a string
+            .stderr: stderr of the program, a string
+
+        """
+        # We make a temp directory to serve as the home of the sandboxed code.
+        # It has a writable "tmp" directory within it for temp files.
+        with temp_directory() as homedir:
+
+            # Make directory readable by other users ('sandbox' user needs to be
+            # able to read it).
+            os.chmod(homedir, 0775)
+
+            # Make a subdir to use for temp files, world-writable so that the
+            # sandbox user can write to it.
+            tmptmp = os.path.join(homedir, "tmp")
+            os.mkdir(tmptmp)
+            os.chmod(tmptmp, 0777)
+
+            argv = argv or []
+            env = {'TMPDIR': 'tmp'}
+
+            # All the supporting files are copied into our directory.
+            for filename in files or ():
+                dest = os.path.join(homedir, os.path.basename(filename))
+                if os.path.islink(filename):
```

Comment:
We need to support symlinks among the files we copy into the sandbox.  If the directory to copy contains files linked to each other, then the copy needs to have files linked to each other.